go/libraries/doltcore/sqle/rowexec/lookup_join.go

```diff
@@ -0,0 +1,327 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rowexec
+
+import (
+	"context"
+	"io"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/index"
+	"github.com/dolthub/dolt/go/store/pool"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+func rowIterTableLookupJoin(
+	srcIter prolly.MapIter,
+	dstIter index.SecondaryLookupIter,
+	mapping *lookupMapping,
+	rowJoiner *prollyToSqlJoiner,
+	srcFilter, dstFilter, joinFilter sql.Expression,
+	isLeftJoin bool,
+	excludeNulls bool,
+) (sql.RowIter, error) {
+	return newLookupKvIter(srcIter, dstIter, mapping, rowJoiner, srcFilter, dstFilter, joinFilter, isLeftJoin, excludeNulls)
+}
+
+type lookupJoinKvIter struct {
+	srcIter prolly.MapIter
+	srcKey  val.Tuple
+	srcVal  val.Tuple
+
+	dstIter index.SecondaryLookupIter
+	dstKey  val.Tuple
+
+	// keyTupleMapper inputs (srcKey, srcVal) to create a dstKey
+	keyTupleMapper *lookupMapping
+
+	// projections
+	joiner *prollyToSqlJoiner
+
+	// TODO convert sql.Expression to static prolly expressions
```

Comment:
Resolve this TODO

Reply:
I think this is out of scope. The two observations are (1) reimplementing every `sql.Expression.Eval` in the `prolly` package is a lift. And (2) if the coster picked a lookup join in spite of this hanging filter, that usually means whatever we did push into the left is already restrictive. Obviously there are edge cases, but a large left side usually favors hash or merge joins. This is all hand-wavy, but I think 2x throughput on big lookup joins will be a win most of the time. And if someone hits this, the short-term fix of hinting hash join is an easy fix while we implement prolly side filters. If this doesn't track let me know, I do think this is the one downside of this PR.

Reply:
I think this is fine, but this ought to be more spelled out in the comments. Leaving it as just a TODO makes it feel like the PR is unfinished.